### PR TITLE
DOCS-11231 hashing client credentials

### PR DESCRIPTION
**Pull request by priyaayyar**
Created a new topic for hashing client credentials

### CODE REVIEW

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 1]
```diff
@@ -0,0 +1,40 @@
+= Configure Credentials Hashing for Mule 4
+ifndef::env-site,env-github[]
+include::_attributes.adoc[]
+endif::[]
+:keywords: hashing, client credentials, auto-generated, raml, wsdl, apikit, http
+
+You enable hashing of your client credentials to add another layer of security for your APIs by hiding the client secrets from the Mule runtime engine (Mule) server. When you enable client hashing, API gateway stores the client credentials in memory, with only the client secret part hashed.
```
Comment (kbostrom) [inline modules/ROOT/pages/hash-client-credentials.adoc]:
```suggestion
Enable hashing of your client credentials to add another layer of security for your APIs by hiding the client secrets from the Mule runtime engine (Mule) server. When you enable client hashing, API gateway stores the client credentials in memory, with only the client secret hashed. 
```

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ifndef::env-site,env-github[]
+include::_attributes.adoc[]
+endif::[]
+:keywords: hashing, client credentials, auto-generated, raml, wsdl, apikit, http
+
+You enable hashing of your client credentials to add another layer of security for your APIs by hiding the client secrets from the Mule runtime engine (Mule) server. When you enable client hashing, API gateway stores the client credentials in memory, with only the client secret part hashed. 
+
+Additionally, if you enabled both encryption and hashing in API gateway, the contract information in memory is hashed, but not encrypted. When client hashing (disabled by default) is enabled, it uses the SHA-256 hashing algorithm.
+
+If you modify the configuration to enable or disable API gateway for hashing and restart your Mule server for the updated configuration to apply, your incoming requests are not affected and they continue to be processed normally.
```
Comment (kbostrom) [inline modules/ROOT/pages/hash-client-credentials.adoc]:
```suggestion
If you modify the configuration to enable or disable hashing in API gateway and then restart your Mule server to apply the updated configuration, your incoming requests are not affected and they continue to be processed normally. 
```

Comment (kbostrom) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Question: if restarting applies the update, why does that not affect processing behavior from that point?

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+You enable hashing of your client credentials to add another layer of security for your APIs by hiding the client secrets from the Mule runtime engine (Mule) server. When you enable client hashing, API gateway stores the client credentials in memory, with only the client secret part hashed. 
+
+Additionally, if you enabled both encryption and hashing in API gateway, the contract information in memory is hashed, but not encrypted. When client hashing (disabled by default) is enabled, it uses the SHA-256 hashing algorithm.
+
+If you modify the configuration to enable or disable API gateway for hashing and restart your Mule server for the updated configuration to apply, your incoming requests are not affected and they continue to be processed normally. 
+
+== How Client Credential Hashing Works
+
+When a request is received, API gateway obtains the client secret from the client credentials of that request and hashes it. API gateway then compares the client ID and the hashed client secret from the request with the credentials stored in memory.
```
Comment (kbostrom) [inline modules/ROOT/pages/hash-client-credentials.adoc]:
```suggestion
With client credential hashing enabled, when a request is received, API gateway obtains the client secret from the client credentials of that request and hashes it. API gateway then compares the client ID and the hashed client secret from the request with the credentials stored in memory. 
```

Comment (kbostrom) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Should we add "...then compares **and verifies** the client ID and the hashed client secret from the request with the credentials stored in memory **before passing the request to Mule**."

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+When a request is received, API gateway obtains the client secret from the client credentials of that request and hashes it. API gateway then compares the client ID and the hashed client secret from the request with the credentials stored in memory. 
+
+== Configure Client Credential Hashing
+
+To enable client hashing, start the Mule server with the `anypoint.platform.hash_clients` property enabled. You can enable the property by either running the start command or by configuring the property in the `wrapper.conf` file in the conf directory:
+
+* start command: 
++
+`$MULE_HOME/bin/mule start -M-Danypoint.platform.hash_clients=true`
+*wrapper.conf property:
```
Comment (kbostrom) [inline modules/ROOT/pages/hash-client-credentials.adoc]:
```suggestion
* `wrapper.conf` property:
```

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+If you modify the configuration to enable or disable API gateway for hashing and restart your Mule server for the updated configuration to apply, your incoming requests are not affected and they continue to be processed normally. 
+
+== How Client Credential Hashing Works
+
+When a request is received, API gateway obtains the client secret from the client credentials of that request and hashes it. API gateway then compares the client ID and the hashed client secret from the request with the credentials stored in memory. 
+
+== Configure Client Credential Hashing
+
+To enable client hashing, start the Mule server with the `anypoint.platform.hash_clients` property enabled. You can enable the property by either running the start command or by configuring the property in the `wrapper.conf` file in the conf directory:
```
Comment (kbostrom) [inline modules/ROOT/pages/hash-client-credentials.adoc]:
```suggestion
To enable client hashing, start the Mule server with the `anypoint.platform.hash_clients` property enabled. You can enable the property by either running the `start` command or by configuring the property in the `wrapper.conf` file in the `conf` directory:
```

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+* start command:
```
Comment (kbostrom) [inline modules/ROOT/pages/hash-client-credentials.adoc]:
```suggestion
* `start` command: 
```

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+To enable client hashing, start the Mule server with the `anypoint.platform.hash_clients` property enabled. You can enable the property by either running the start command or by configuring the property in the `wrapper.conf` file in the conf directory:
+
+* start command: 
++
+`$MULE_HOME/bin/mule start -M-Danypoint.platform.hash_clients=true`
+*wrapper.conf property:
++
+`wrapper.java.additional.<n>=-Danypoint.platform.hash_clients=true`
+
+If you add `wrapper.java.additional.n` entries to the configuration file, change each instance of `<n>` to a consecutive number. Otherwise, Java does not parse the properties correctly.
```
Comment (kbostrom) [inline modules/ROOT/pages/hash-client-credentials.adoc]: What do we mean by "a consecutive number"? 
Consecutive to what beginning number?
Does the numbering begin with 1 and then increase by 1 for each consecutive `n` instance?
Clarify.

**Review by kbostrom — DISMISSED** [review]
CE

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 1 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 2 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 3 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 4 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 5 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 6 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 7 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+Additionally, if you enabled both encryption and hashing in API gateway, the contract information in memory is hashed, but not encrypted. When client hashing (disabled by default) is enabled, it uses the SHA-256 hashing algorithm.
+
+If you modify the configuration to enable or disable hashing in API gateway and then restart your Mule server to apply the updated configuration, your incoming requests are not affected and they continue to be processed normally. 
+
+== How Client Credential Hashing Works
+
+When a request is received, API gateway obtains the client secret from the client credentials of that request and hashes it. API gateway then compares the client ID and the hashed client secret from the request with the credentials stored in memory. 
+
+With client credential hashing enabled, when a request is received, API gateway obtains the client secret from the client credentials of that request and hashes it. API gateway then compares and verifies the client ID and the hashed client secret from the request with the credentials stored in memory before passing the request to Mule.
```
Comment (Pgarrett) [inline modules/ROOT/pages/hash-client-credentials.adoc:15]: This paragraph is almost a duplicate of the one above

File: modules/ROOT/pages/hash-client-credentials.adoc [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+To enable client hashing, start the Mule server with the `anypoint.platform.hash_clients` property enabled. You can enable the property by either running the `start` command or by configuring the property in the `wrapper.conf` file in the `conf` directory:
+
+* `start` command: 
++
+`$MULE_HOME/bin/mule start -M-Danypoint.platform.hash_clients=true`
+* `wrapper.conf` property:
++
+`wrapper.java.additional.<n>=-Danypoint.platform.hash_clients=true`
+
+If you add `wrapper.java.additional.n` entries to the configuration file, change each instance of `<n>` to a consecutive number. Otherwise, Java does not parse the properties correctly. For example:
```
Comment (Pgarrett) [inline modules/ROOT/pages/hash-client-credentials.adoc]: Add `<n>` to `wrapper.java.additional.n`

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 8 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc:15]: Good catch! rewrote this para and forgot to delete the old one. :-)

File: modules/ROOT/pages/hash-client-credentials.adoc (hunk 9 above)
Comment (priyaayyar) [reply modules/ROOT/pages/hash-client-credentials.adoc]: Done! Thanks!

**Review by Pgarrett — APPROVED** [review]

**Review by kbostrom — APPROVED** [review]
ra

**Pull request merged**